active/0051-packaging.md

```diff
@@ -0,0 +1,338 @@
+reader to understand.
+
+## Packaging
+
+It gives you granular control over the final build output. It could be used in many
+different ways (we are going to go over use cases below). Note, it isn't meant to be
+used for "postprocess" transformations; "postprocess" is called after packaging is
+finished.
+
+Currently, Ember.js application and all of its depedencies get assembled under one
+directory with the following structure:
+
+```ruby
+bundler:js:input/
+├── addon-tree-output/
+├── the-app-name-folder/
+├── node_modules/
+└── vendor/
+```
+
+where:
+
++ `addon-tree-output` is a folder that contains dependencies from Ember add-ons
++ `the-app-name-folder` is a folder that contains Ember application code
++ `node_modules` is a folder that contains node dependencies
++ `vendor` is a folder that contains other dependencies.
+
+Note, for clarity purposes we should rename `addon-tree-output` to `addon-modules` as
+both `tree` and `output` don't communicate well about the contents of the folder.
+
+During packaging process the final output will be generated (everything that currently
+resides under `dist/` folder when a developer runs `ember build`).
+
+### `package` API
+
+A new public `package` method will be introduced to open up a way to customise packaging process:
+
+```javascript
+// ember-cli-build.js
+const EmberApp = require('ember-cli/lib/broccoli/ember-app');
+
+module.exports = function(defaults) {
+  const app = new EmberApp(defaults, {
+    package(inputTree) {
+      // customise `inputTree`
+      // and return customised `inputTree`
+    }
+  });
+
+  return app.toTree();
+}
+```
+
+`package` function has the following signature:
+
+```typescript
+interface EmberApp {
+  package(inputTree: BroccoliTree): BroccoliTree;
+}
+```
+
+where `inputTree` will have the following structure:
+
+```ruby
+bundler:js:input/
+├── addon-modules/
+├── the-app-name-folder/
+├── node_modules/
+└── vendor/
+```
+
+Note, that `package` method must return a broccoli tree.
+
+This change should be behind an experiment flag, `PACKAGING`.
+This will allow us to start experimenting right away and not being
+tied to a particular release cycle.
+
+Note, that `package` is optional. If you don't define it, you're
+effectively "opting out" of the feature and using the default
+behaviour.
+
+### `packagerFor` API
+
+It's important to make it easy for users to still use default Ember CLI
+packaging.
+
+`packagerFor` is a way for the users to access out-of-the-box packaging while
+still be able to customise the final build output.
+
+```javascript
+// ember-cli-build.js
+const EmberApp = require('ember-cli/lib/broccoli/ember-app');
+const packagerFor = require('ember-cli/lib/broccoli/packager-for');
+
+module.exports = function(defaults) {
+  const app = new EmberApp(defaults, {
+    package(inputTree) {
+      const defaultPackager = packagerFor(app);
+      // customise `inputTree`
+
+      return defaultPackager(inputTree);
```

Comment:
Assuming the `package` function receives a precompiled tree, if someone wanted to do custom bundling for their app, would they be opting out of our compilation system? Would they be responsible for doing their own compilation, even though they didn't want to affect the compilation step?